server/mqtt.go

```diff
@@ -2481,15 +2482,55 @@ func (as *mqttAccountSessionManager) loadRetainedMessagesForSubject(rms map[stri
 			log.Warnf("failed to load retained message for subject %q: %v", loadSubject, err)
 			continue
 		}
-		var rm mqttRetainedMsg
-		if err := json.Unmarshal(jsm.Data, &rm); err != nil {
+		rm, err := mqttDecodeRetainedMessage(jsm.Header, jsm.Data)
+		if err != nil {
 			log.Warnf("failed to decode retained message for subject %q: %v", loadSubject, err)
 			continue
 		}
-		rms[subject] = &rm
+		rms[subject] = rm
 	}
 }
 
+// Composes a NATS message for a storeable mqttRetainedMsg.
+func mqttEncodeRetainedMessage(rm *mqttRetainedMsg) (natsMsg []byte, headerLen int) {
+	// Overallocate, 2K for all the header, JSON overhead
+	buf := bytes.NewBuffer(make([]byte, 0,
+		len(rm.Msg)+len(rm.Subject)+len(rm.Origin)+len(rm.Topic)+len(rm.Source)+2*1024))
```

Comment:
Seems a bit excessive, especially that the buffer can grow if needed.

Reply:
You're right, the "overhead" is fixed already; adjusted it down to ~140 bytes; still too much but at this point doesn't matter, no need to be overly precise.